# 01. Import libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 02. Import data

In [2]:
# Create path
path = r'C:\Users\opetr\Documents\Instacart Basket Analysis'

In [3]:
# Import ords_prods_merge
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_new.pkl'))

# 03. TASK 4.8

In [4]:
# Group product_name column
ords_prods_merge.groupby('product_name')

In [5]:
# Perform a single aggregation
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [6]:
# Use the mean() function to achieve the same result
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [ ]:
#3. Analyze the result. How do the results for the entire dataframe differ from those of the subset?

#### The results for the entire dataframe show a comprehensive list of all department ids. Additionally, the average from the subset it different from the average when looking at the entire dataframe. E.g. dept 4 had an average of 18.83 when looking at the first million rows. When looking at the entire dataframe, the same department has an avearge of 17.81

In [7]:
#4. Create a loyalty flag for existing customers using the transform() and loc() functions
# Create a max_order column
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [8]:
# Change head printing max
pd.options.display.max_rows = None

In [9]:
# Check the head
ords_prods_merge.head(100)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0              0   2539329        1             1                   2   
1              1   2398795        1             2                   3   
2              2    473747        1             3                   3   
3              3   2254736        1             4                   4   
4              4    431534        1             5                   4   
5              5   3367565        1             6                   2   
6              6    550135        1             7                   1   
7              7   3108588        1             8                   1   
8              8   2295261        1             9                   1   
9              9   2550362        1            10                   4   
10           144   2968173       15            15                   1   
11           146   1870022       15            17                   2   
12           147   1911383       15            18                   2   
13           150   2715276       15            21                   1   
14           151    487368       15            22                   1   
15           209   2293453       19             2                   5   
16           212   1973799       19             5                   6   
17           214    532817       19             7                   4   
18           232   1573906       21            10                   3   
19           458   1593000       31            10                   3   
20           465   2231262       31            17                   3   
21           617   2580647       43             6                   4   
22           620   2187180       43             9                   4   
23           751   2497897       52             1                   1   
24           752   1318871       52             2                   1   
25           753   1261384       52             3                   1   
26           755    580568       52             5                   2   
27           757   2428073       52             7                   2   
28           759    342306       52             9                   1   
29           760   1520353       52            10                   2   
30           761   2542086       52            11                   2   
31           762    944694       52            12                   1   
32           765   1268191       52            15                   2   
33           767   1122089       52            17                   2   
34           768   1498922       52            18                   2   
35           769    180919       52            19                   3   
36           771   2695875       52            21                   5   
37          1022   3205960       67             1                   3   
38          1023    970922       67             2                   1   
39          1024    596585       67             3                   1   
40          1025   2037087       67             4                   2   
41          1026   2474111       67             5                   3   
42          1027   1568481       67             6                   3   
43          1029   2048527       67             8                   2   
44          1030   2389031       67             9                   2   
45          1031   2378562       67            10                   1   
46          1032   1515698       67            11                   1   
47          1033   1928050       67            12                   2   
48          1034   1949274       67            13                   1   
49          1035    551869       67            14                   1   
50          1037    768572       67            16                   4   
51          1039    150528       67            18                   5   
52          1040     29758       67            19                   0   
53          1042   1558976       67            21             

In [10]:
# Create condition 1
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [11]:
# Create condition 2
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [12]:
# Create condition 3
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [13]:
# Check frequency
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [14]:
# Return the first 60 rows that only include 3 columns
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

In [16]:
#5. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer)

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

#### This shows that there is a difference in the prices of products purchased by loyal customers vs regular or new customers. Loyal customers spend, on average, the least amount of money on products. 

In [17]:
#6. Create a spending flag for each user based on the average price across all their orders
# Group
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [18]:
# Transform
ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [19]:
# Check output
ords_prods_merge.head(20)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0              0   2539329        1             1                   2   
1              1   2398795        1             2                   3   
2              2    473747        1             3                   3   
3              3   2254736        1             4                   4   
4              4    431534        1             5                   4   
5              5   3367565        1             6                   2   
6              6    550135        1             7                   1   
7              7   3108588        1             8                   1   
8              8   2295261        1             9                   1   
9              9   2550362        1            10                   4   
10           144   2968173       15            15                   1   
11           146   1870022       15            17                   2   
12           147   1911383       15            18                   2   
13           150   2715276       15            21                   1   
14           151    487368       15            22                   1   
15           209   2293453       19             2                   5   
16           212   1973799       19             5                   6   
17           214    532817       19             7                   4   
18           232   1573906       21            10                   3   
19           458   1593000       31            10                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   
15                 14                     6.0         196                  3   
16                 12                     8.0         196                 15   
17                 17                     6.0         196                  1   
18                 10                     6.0         196                  2   
19                  8                     7.0         196                 17   

    reordered  ...  department_id prices  _merge    price_range_loc  \
0           0  ...              7    9.0    both  Mid-range product   
1           1  ...              7    9.0    both  Mid-range product   
2           1  ...              7    9.0    both  Mid-range product   
3           1  ...              7    9.0    both  Mid-range product   
4           1  ...              7    9.0    both  Mid-range product   
5           1  ...              7    9.0    both  Mid-range product   
6           1  ...              7    9.0    both  Mid-range product   
7           1  ...              7    9.0    both  Mid-range product   
8           1  ...              7    9.0    both  Mid-range product   
9           1  ...              7    9.0    both  Mid-range product   
10      

In [20]:
# Create condition 1
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [22]:
# Create condition 2
ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [23]:
# Check frequency
ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

In [24]:
# Return the first 25 rows
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(25)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2

In [25]:
#7. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column
# Transform
ords_prods_merge['median_order_frequency']= ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [26]:
# Check output
ords_prods_merge.head(20)

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0              0   2539329        1             1                   2   
1              1   2398795        1             2                   3   
2              2    473747        1             3                   3   
3              3   2254736        1             4                   4   
4              4    431534        1             5                   4   
5              5   3367565        1             6                   2   
6              6    550135        1             7                   1   
7              7   3108588        1             8                   1   
8              8   2295261        1             9                   1   
9              9   2550362        1            10                   4   
10           144   2968173       15            15                   1   
11           146   1870022       15            17                   2   
12           147   1911383       15            18                   2   
13           150   2715276       15            21                   1   
14           151    487368       15            22                   1   
15           209   2293453       19             2                   5   
16           212   1973799       19             5                   6   
17           214    532817       19             7                   4   
18           232   1573906       21            10                   3   
19           458   1593000       31            10                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   
15                 14                     6.0         196                  3   
16                 12                     8.0         196                 15   
17                 17                     6.0         196                  1   
18                 10                     6.0         196                  2   
19                  8                     7.0         196                 17   

    reordered  ...  _merge    price_range_loc     busiest_day    busiest_days  \
0           0  ...    both  Mid-range product  Regularly busy  Regularly busy   
1           1  ...    both  Mid-range product  Regularly busy    Slowest days   
2           1  ...    both  Mid-range product  Regularly busy    Slowest days   
3           1  ...    both  Mid-range product      Least busy    Slowest days   
4           1  ...    both  Mid-range product      Least busy    Slowest days   
5           1  ...    both  Mid-range product  Regularly busy  Regularly busy   
6           1  ...    both  Mid-range product  Regularly busy    Busiest days   
7           1  ...    both  Mid-range product  Regularly busy    Busiest days   
8           1  ...    both  Mid-range product  Regularly bus

In [27]:
# Create condition 1
ords_prods_merge.loc[ords_prods_merge['median_order_frequency'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [28]:
# Create condition 2
ords_prods_merge.loc[(ords_prods_merge['median_order_frequency'] > 10) & (ords_prods_merge['median_order_frequency'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [29]:
# Create condition 2
ords_prods_merge.loc[ords_prods_merge['median_order_frequency'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [30]:
# Check frequency
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [31]:
# Return the first 25 rows
ords_prods_merge[['user_id', 'order_frequency_flag', 'order_number']].head(25)

user_id   order_frequency_flag  order_number
0         1  Non-frequent customer             1
1         1  Non-frequent customer             2
2         1  Non-frequent customer             3
3         1  Non-frequent customer             4
4         1  Non-frequent customer             5
5         1  Non-frequent customer             6
6         1  Non-frequent customer             7
7         1  Non-frequent customer             8
8         1  Non-frequent customer             9
9         1  Non-frequent customer            10
10       15      Frequent customer            15
11       15      Frequent customer            17
12       15      Frequent customer            18
13       15      Frequent customer            21
14       15      Frequent customer            22
15       19      Frequent customer             2
16       19      Frequent customer             5
17       19      Frequent customer             7
18       21      Frequent customer            10
19       31      Frequent customer            10
20       31      Frequent customer            17
21       43      Frequent customer             6
22       43      Frequent customer             9
23       52      Frequent customer             1
24       52      Frequent customer             2

In [32]:
#9. Export your dataframe as a pickle file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_grouped.pkl'))